terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,116 @@
 package provider
 
 import (
-	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"context"
+	"fmt"
 
+	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"github.com/gravitational/trace"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/gravitational/teleport/api/client"
+	"github.com/gravitational/teleport/api/types"
 )
 
+// teleport_role resource definition
 func resourceTeleportRole() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceTeleportRoleUpsert,
-		Read:   resourceTeleportRoleRead,
-		Update: resourceTeleportRoleUpsert,
-		Delete: resourceTeleportRoleDelete,
-		Exists: resourceTeleportRoleExists,
+		CreateContext: resourceRoleCreate,
+		ReadContext:   resourceRoleRead,
+		UpdateContext: resourceRoleUpdate,
+		DeleteContext: resourceRoleDelete,
 
 		Schema: tfschema.SchemaRoleV3(),
 	}
 }
 
-func resourceTeleportRoleUpsert(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*client.Client)
-	name := d.Get("name").(string)
+// resourceRoleCreate creates Teleport role
+func resourceRoleCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c := m.(*client.Client)
 
-	// roleSpec := services.RoleSpecV3{}
+	name, err := getRoleName(d)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// role, err := services.NewRole(name, roleSpec)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	role, err := types.NewRole(name, types.RoleSpecV3{})
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	roleV3 := role.(*types.RoleV3)
+	tfschema.UnmarshalRoleV3(d, roleV3)
 
-	// FIXME: Client.CreateRole() is not implemented yet in api/client
-	// err = client.CreateRole(role)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	err = c.UpsertRole(ctx, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	d.SetId(name)
-	return nil
+	d.SetId(roleV3.GetName())
+
+	resourceRoleRead(ctx, d, m)
```

Comment:
check for errors here and below

Reply:
GetName()/SetId() do not return errors. Replaced `resourceRoleRead()` with `return resourceRoleRead()` here and below.